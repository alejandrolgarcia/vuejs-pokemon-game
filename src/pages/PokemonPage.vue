<template>
    <h1 v-if="!pokemon">Cargando pokemón...</h1>
    <id v-else>
        <h1>¿Quién es este pokemón?</h1>
        <!-- Img pokemon -->
        <PokemonImage 
            :pokemon-id="pokemon.id" 
            :show-pokemon="showPokemon" />

        <PokemonOptions 
            :pokemons="pokemonArray" 
            @selection="checkAnswer($event)" />

        <div v-if="showAnswer">
            <h2>{{ message }}</h2>
            <button @click="newGame" class="primary small">
                Nuevo Juego
            </button>
        </div>
    </id>
</template>

<script>
import PokemonImage from '@/components/PokemonImage.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        PokemonImage,
        PokemonOptions,
    },

    data() {
        return {
            pokemonArray: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },

    methods: {
        async mixPokemonsArray() {
            this.pokemonArray = await getPokemonOptions()

            const rndInt = Math.floor( Math.random() * 4 )
            this.pokemon = this.pokemonArray[ rndInt ]
        },
        checkAnswer( selectedId ) {

            this.showPokemon = true
            this.showAnswer  = true

            if ( selectedId === this.pokemon.id ) {
                this.message = `Correcto, es ${ this.pokemon.name }!`
            } else {
                this.message = `Oops, era ${ this.pokemon.name } :'(`
            }
        },
        newGame() {
            this.showPokemon  = false,
            this.showAnswer   = false,
            this.pokemonArray = [],
            this.pokemon      = null,
            this.mixPokemonsArray()
        }
    },
    // Lifecycle hooks
    mounted() {
        this.mixPokemonsArray()
    },

}
</script>

<style scoped>
    button {
        -moz-appearance: none;
        -moz-transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out;
        -ms-appearance: none;
        -ms-transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out;
        -webkit-appearance: none;
        -webkit-transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out;
        appearance: none;
        background-color: transparent;
        border-radius: 0.375em;
        border: 0;
        box-shadow: inset 0 0 0 2px #f56a6a;
        color: #f56a6a !important;
        cursor: pointer;
        display: inline-block;
        font-size: 0.8em;
        font-weight: 700;
        height: 3.5em;
        letter-spacing: 0.075em;
        line-height: 3.5em;
        margin: 0 5px;
        padding: 0 2.25em;
        text-align: center;
        text-decoration: none;
        text-transform: uppercase;
        transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out;
        white-space: nowrap;
    }
    button.primary {
        background-color: #f56a6a;
        box-shadow: none;
        color: #ffffff !important;
    }
    button:hover {
        background-color: rgba(245, 106, 106, 0.05);
    }
    button.primary:hover {
        background-color: #f67878; 
    }

    button.small {
        font-size: 0.6em;
    }
</style>
