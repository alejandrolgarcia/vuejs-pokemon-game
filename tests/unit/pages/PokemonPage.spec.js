import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {
    
    let wrapper

    beforeEach( () => {
        wrapper = shallowMount( PokemonPage )
    })

    test('debe hacer match con el snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe de llamar mixPokemonArray al montar el componente', () => {
        
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonsArray' )
        const wrapper = shallowMount( PokemonPage )

        expect( mixPokemonArraySpy ).toHaveBeenCalled()

    })

    test('debe hacer match con el snapshot cuando carga el pokemon', () => {
        
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArray: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    points: 0,
                    lives: 3
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()

    })

    test('debe mostrar los componentes de PokemonImage y PokemonOptions', () => {

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArray: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    points: 0,
                    lives: 3
                }
            }
        })

        const pokemonImage = wrapper.find('pokemon-image-stub')
        const pokemonOptions = wrapper.find('pokemon-options-stub')
        const pokemonScore = wrapper.find('pokemon-score-stub')

        expect( pokemonImage.exists() ).toBeTruthy()
        expect( pokemonOptions.exists() ).toBeTruthy()
        expect( pokemonScore.exists() ).toBeTruthy()

        expect( pokemonImage.attributes('pokemonid') ).toBe('15')
        expect( pokemonOptions.attributes('pokemons') ).toBeDefined()
        
    })

    test('pruebas en checkAnswer', async() => {

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArray: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    points: 0,
                    lives: 3
                }
            }
        })

        /**
         * con .vm tenemos acceso al virtual doom de Vue
         */
        await wrapper.vm.checkAnswer( 15 )

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe( true )
        expect( wrapper.find('h2').text() ).toBe( `Correcto, es ${ pokemons[0].name }!` )

        await wrapper.vm.checkAnswer( 20 )
        expect( wrapper.vm.message ).toBe(`Oops, era ${ pokemons[0].name } :'(`)


    })

})
