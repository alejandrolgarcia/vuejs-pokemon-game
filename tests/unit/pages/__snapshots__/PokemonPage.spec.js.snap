// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe hacer match con el snapshot 1`] = `<h1>Cargando pokemón...</h1>`;

exports[`PokemonPage Component debe hacer match con el snapshot cuando carga el pokemon 1`] = `
<id>
  <h1>¿Quién es este pokemón?</h1><!-- Img pokemon -->
  <pokemon-image-stub pokemonid="15" showpokemon="false"></pokemon-image-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <pokemon-score-stub points="0" lives="3"></pokemon-score-stub>
  <!--v-if-->
</id>
`;
